packages/edit-navigation/src/store/actions.js

```diff
@@ -49,171 +39,380 @@ export function setSelectedMenuId( menuId ) {
+ * @param {string}   entityType Entity type.
+ * @param {Object[]} changeset  The changeset.
+ * @param {Object[]} results    The results of persisting the changeset.
+ * @return {Object[]} A list of failed changeset entries.
+ */
+const getFailedChanges = ( kind, entityType, changeset, results ) => ( {
+	registry,
+} ) => {
+	const failedDeletes = zip( changeset, results )
+		.filter(
+			( [ change, result ] ) =>
+				change.type === 'delete' &&
+				! result?.hasOwnProperty( 'deleted' )
 		)
+		.map( ( [ change ] ) => change );
+
+	const failedUpdates = changeset.filter(
+		( change ) =>
+			change.type === 'update' &&
+			change.id &&
+			registry
+				.select( coreDataStore )
+				.getLastEntitySaveError( kind, entityType, change.id )
 	);
 
-	return yield apiFetch( {
-		url: ajaxurl || '/wp-admin/admin-ajax.php',
-		method: 'POST',
-		body,
-	} );
+	return [ ...failedDeletes, ...failedUpdates ];
+};
+
+/**
+ * Diffs oldEntityRecords and desiredEntityRecords, returning a list of
+ * create, delete, and update tasks necessary to go from the former to the latter.
+ *
+ * @param {string}   kind                 Entity kind.
+ * @param {string}   type                 Entity type.
+ * @param {Object[]} oldEntityRecords     The entity records that are currently persisted.
+ * @param {Object[]} desiredEntityRecords The entity records are to be persisted.
+ * @return {Function} An action creator
+ */
+const prepareChangeset = (
+	kind,
+	type,
+	oldEntityRecords,
+	desiredEntityRecords
+) => ( { registry } ) => {
+	const deletedEntityRecordsIds = new Set(
+		diff(
+			oldEntityRecords.map( ( { id } ) => id ),
+			desiredEntityRecords.map( ( { id } ) => id )
+		)
+	);
+
+	const changes = [];
+	// Enqueue updates
+	for ( const entityRecord of desiredEntityRecords ) {
+		if (
+			! entityRecord?.id ||
+			deletedEntityRecordsIds.has( entityRecord?.id )
+		) {
+			continue;
+		}
+
+		// Update an existing entity record.
+		registry
+			.dispatch( coreDataStore )
+			.editEntityRecord( kind, type, entityRecord.id, entityRecord, {
+				undoIgnore: true,
+			} );
+
+		const hasEdits = registry
+			.select( coreDataStore )
+			.hasEditsForEntityRecord( kind, type, entityRecord.id );
+
+		if ( ! hasEdits ) {
+			continue;
+		}
+
+		changes.unshift( {
+			type: 'update',
+			id: entityRecord.id,
+			batchTask: ( { saveEditedEntityRecord } ) =>
+				saveEditedEntityRecord( kind, type, entityRecord.id ),
+		} );
+	}
+
+	// Enqueue deletes
+	for ( const entityRecordId of deletedEntityRecordsIds ) {
+		changes.unshift( {
+			type: 'delete',
+			id: entityRecordId,
+			batchTask: ( { deleteEntityRecord } ) =>
+				deleteEntityRecord( kind, type, entityRecordId, {
+					force: true,
+				} ),
+		} );
+	}
+
+	return changes;
+};
+
+/**
+ * Returns elements of list A that are not in list B.
+ *
+ * @param {Array} listA List A.
+ * @param {Array} listB List B.
+ * @return {Array} elements of list A that are not in list B.
+ */
+function diff( listA, listB ) {
```

Comment:
Can we use `_.difference`? WP Admin always includes lodash so making use of it where possible reduces surface area for bugs and number of bytes sent to the browser.

Reply:
good idea!